torch/nested/_nestedtensor.py

```diff
@@ -0,0 +1,118 @@
+import torch
+import numbers
+
+
+def _filter_impl(args, kwargs):
+    if kwargs is None:
+        kwargs = {}
+    impl_args = []
+    for a in args:
+        if isinstance(a, torch.NestedTensor):
+            impl_args.append(a._impl)
+        elif torch.is_tensor(a):
+            impl_args.append(a)
+        elif isinstance(a, list):
+            a_impl, _ = _filter_impl(a, {})
+            impl_args.append(a_impl)
+        elif isinstance(a, tuple):
+            a_impl, _ = _filter_impl(a, {})
+            impl_args.append(tuple(a_impl))
+        else:
+            impl_args.append(a)
+    impl_kwargs = {
+        k: v._impl if isinstance(v, torch.NestedTensor) else v
+        for (k, v) in kwargs.items()
+    }
+    return impl_args, impl_kwargs
+
+
+def _wrap_result(result):
+    if isinstance(result, list):
+        return list(_wrap_result(r) for r in result)
+    if isinstance(result, tuple):
+        return tuple(_wrap_result(r) for r in result)
+    return (
+        torch.NestedTensor(result)
+        if torch.is_tensor(result) and torch.is_nt_impl(result)
+        else result
+    )
+
+
+class NestedTensor:
```

Comment:
Need more docs...

Reply:
Can I add docs in a follow up?

Reply:
Also need an explanation of why we are wrapping NestedTensor in this way

Reply:
Because it doesn't have a size. So, something as simple as printing doesn't work. We should migrate this to torch_dispatch as a follow-up.

Reply:
I think the way of fixing printing which is more in line with how, e.g., sparse is currently done, is to just override the tensor printer to check for nested tensor (so you'll need to bind an `is_nested_tensor` check somehow; current prevailing style is to do it as a method but TBH for experimental stuff it's better to cook it as a function first). You delegate size call to the inner implementation so removing the wrapper is no worse...

Reply:
Yes, now that we're in tree that is feasible. Out of tree that is impossible. That's why this is written this way.

Reply:
OK I filed an issue for printing specifically at https://github.com/pytorch/pytorch/issues/73506 I'm not sure what you want to do about serialization though: because it's a C++ struct difference the logical interposition point would be in C++, but pickling is a purely Python side idea (TorchScript C++ pickler/unpickler notwithstanding), so you've gotta put the logic somewhere.